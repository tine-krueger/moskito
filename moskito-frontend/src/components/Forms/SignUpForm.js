import styled from 'styled-components/macro'
import ButtonBackGroup from "../Button/ButtonBackGroup"
import InputField from './FormElements/InputField'
import useForm from '../../hooks/useForm'
import Redirection from './FormElements/Redirection'

export default function SignUpForm({
    loginErrors, 
    isRegistered, 
    userRegistration
}) {

    const { inputs, isPasswordEqual, handleChange, handleSubmit, handleClick } = useForm({
        firstName: '',
        lastName:'',
        email:'',
        password:'',
        passwordControl:''
    }, signup)
    

    return (
        <SigninFormStyled onSubmit={handleSubmit}>
            {isRegistered && <Redirection/>}
            {loginErrors && loginErrors.map((error, index) => <RedParagraph key={index}>{error}</RedParagraph>)}

            <InputField type='text' name='firstName' value={inputs.firstName} onChange={handleChange} placeholder={'Vorname'} marginBottom={'sm'}/>
            <InputField type='text' name='lastName' value={inputs.lastName} onChange={handleChange} placeholder={'Nachname'} marginBottom={'sm'}/>
            <InputField type='text' name='email' value={inputs.email} onChange={handleChange} placeholder={'E-Mail'} marginBottom={'sm'}/>

            {!isPasswordEqual && <RedParagraph>Die Passwörter stimmen nicht überein!</RedParagraph>}
            
            <InputField type={'password'} name='password' value={inputs.password} onChange={handleChange} placeholder={'Passwort'} marginBottom={'sm'}/>
            <InputField type={'password'} name='passwordControl' value={inputs.passwordControl} onChange={handleChange} placeholder={'Passwort Wiederholung'}/>
            <ButtonBackGroup text1={'SignUp'} text2={'Zurück'} onClick={handleClick}/>
        </SigninFormStyled>
    )

    function signup() {
        userRegistration(isPasswordEqual, inputs)
    }
}

const SigninFormStyled = styled.form`
    max-width: 500px;
    display: grid;
    margin:2em;
`
const RedParagraph = styled.p`
    color: var(--link);
`
