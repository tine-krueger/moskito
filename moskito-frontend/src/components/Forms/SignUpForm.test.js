import { render } from '@testing-library/react'
import 'jest-styled-components'
import userEvent from '@testing-library/user-event'
import SignUpForm from './SignUpForm'

describe('Sign Up Form', () => {

    it('renders correctly', () => {
        const { container } = render(<SignUpForm/>)
        expect(container).toMatchSnapshot()
    })

    it('shows password not equal text, if passwords are not the same', () => {
        const { getByPlaceholderText, queryByText } = render(<SignUpForm />)
        userEvent.type(getByPlaceholderText('Passwort'), 'SavePasswort')
        userEvent.type(getByPlaceholderText('Passwort Wiederholung'), 'UnsavePasswort')
        expect(queryByText('Die Passwörter stimmen nicht überein!')).toBeInTheDocument()
        expect(queryByText('Die Passwörter stimmen überein!')).not.toBeInTheDocument()
    })

    it('shows no password not equal text, if passwords are the same', () => {
        const { getByPlaceholderText, queryByText } = render(<SignUpForm />)
        userEvent.type(getByPlaceholderText('Passwort'), 'SavePasswort')
        userEvent.type(getByPlaceholderText('Passwort Wiederholung'), 'SavePasswort')
        expect(queryByText('Die Passwörter stimmen nicht überein!')).not.toBeInTheDocument()
    })

    it('shows errors, if errors exist', () => { 
        const mockErrors = ['It has errors', 'It has more errors']
        const { getByText } = render(
            <SignUpForm 
                loginErrors={mockErrors}
            />
        )
        expect(getByText('It has errors')).toBeInTheDocument()
        expect(getByText('It has more errors')).toBeInTheDocument()
    })

    it('redirects if user is registered', () => {
        const mockIsRegistered = true
        const { getByTestId, queryByTestId } = render(
            <SignUpForm 
                isRegistered={mockIsRegistered}
            />
        )
        expect(getByTestId('redirect')).toBeInTheDocument()
        expect(queryByTestId('noredirect')).not.toBeInTheDocument()
    })

    
    it('calls handleSubmit with the right parameters', () => {
        const mockUserRegistration = jest.fn()
        const isPasswordEqual = true
        const { getByPlaceholderText, getByText } = render(
            <SignUpForm 
                userRegistration={mockUserRegistration}
            />
        )
        userEvent.type(getByPlaceholderText('Vorname'), 'Jane')
        userEvent.type(getByPlaceholderText('Nachname'), 'Doe')
        userEvent.type(getByPlaceholderText('E-Mail'), 'jane@doe.de')
        userEvent.type(getByPlaceholderText('Passwort'), 'qwer12!')
        userEvent.type(getByPlaceholderText('Passwort Wiederholung'), 'qwer12!')
        userEvent.click(getByText('SignUp'))
        
        expect(mockUserRegistration).toHaveBeenCalled()
        expect(mockUserRegistration).toHaveBeenCalledWith(
            isPasswordEqual,
            {
                firstName: 'Jane',
                lastName: 'Doe',
                email: 'jane@doe.de',
                password: 'qwer12!',
                passwordControl: 'qwer12!'
            })
    })

})