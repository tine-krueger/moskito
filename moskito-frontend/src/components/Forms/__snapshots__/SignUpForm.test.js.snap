// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up Form renders correctly 1`] = `
.c4 {
  padding: 1em 2em;
  font-family: 'Arvo';
  font-size: 1.1rem;
  border-radius: 10px;
  width: auto;
  border: none;
  background: #c2d6d3;
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-2px -2px 2px 0px rgba(255,255,255,0.5);
  color: #476c85;
  text-transform: uppercase;
  margin: 1em auto 3em auto;
}

.c4:hover {
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3);
  outline: none;
}

.c3 {
  display: grid;
  justify-items: center;
}

.c3 Button {
  margin: .5em;
  width: 10em;
}

.c1 {
  grid-column: 1 / -1;
  width: 100%;
  text-align: center;
  margin: 0 auto 2em;
  margin-bottom: .5em;
}

.c1 h3 {
  text-align: left;
  padding-left: .5em;
  margin: 0 0 1em;
}

.c1 input {
  width: 100%;
  padding: 1em;
  height: 3em;
  font-size: 1em;
  border: none;
  border-radius: 10px;
  color: #6b717e;
  background-color: #d8e6e4;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3);
}

.c1 input:focus {
  outline: none;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3),0 0 0px 1.7px #d3a392;
}

.c2 {
  grid-column: 1 / -1;
  width: 100%;
  text-align: center;
  margin: 0 auto 2em;
  margin-bottom: 1em;
}

.c2 h3 {
  text-align: left;
  padding-left: .5em;
  margin: 0 0 1em;
}

.c2 input {
  width: 100%;
  padding: 1em;
  height: 3em;
  font-size: 1em;
  border: none;
  border-radius: 10px;
  color: #6b717e;
  background-color: #d8e6e4;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3);
}

.c2 input:focus {
  outline: none;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3),0 0 0px 1.7px #d3a392;
}

.c0 {
  max-width: 500px;
  display: grid;
  margin: 2em;
}

<div>
  <form
    class="c0"
  >
    <label
      class="c1"
      data-testid="input-field"
    >
      <h3 />
      <input
        data-testid="firstName"
        name="firstName"
        placeholder="Vorname"
        type="text"
        value=""
      />
    </label>
    <label
      class="c1"
      data-testid="input-field"
    >
      <h3 />
      <input
        data-testid="lastName"
        name="lastName"
        placeholder="Nachname"
        type="text"
        value=""
      />
    </label>
    <label
      class="c1"
      data-testid="input-field"
    >
      <h3 />
      <input
        data-testid="email"
        name="email"
        placeholder="E-Mail"
        type="text"
        value=""
      />
    </label>
    <label
      class="c1"
      data-testid="input-field"
    >
      <h3 />
      <input
        data-testid="password"
        name="password"
        placeholder="Passwort"
        type="password"
        value=""
      />
    </label>
    <label
      class="c2"
      data-testid="input-field"
    >
      <h3 />
      <input
        data-testid="passwordControl"
        name="passwordControl"
        placeholder="Passwort Wiederholung"
        type="password"
        value=""
      />
    </label>
    <div
      class="c3"
    >
      <button
        class="c4"
      >
        SignUp
      </button>
      <button
        class="c4"
      >
        Zur√ºck
      </button>
    </div>
  </form>
</div>
`;
